core/unmarshal_v2.go

```diff
@@ -0,0 +1,521 @@
+	if foundIt && rawSlice != nil {
+		for _, rawMsg := range rawSlice {
+			// 'rawMsg' should contain an instance of the model - we need to unmarshal it into a map.
+			rawMap := make(map[string]json.RawMessage)
+			err = json.Unmarshal(rawMsg, &rawMap)
+			if err != nil {
+				err = fmt.Errorf(errorUnmarshalModel, propInsert(propertyName),
+					reflect.TypeOf(result).Elem().String(), err.Error())
+				return
+			}
+
+			// Reflectively construct a new ptr to a model instance to serve as the receiver of the unmarshal step.
+			rModelValue := reflect.New(reflect.PtrTo(reflect.TypeOf(result).Elem().Elem()))
+
+			// Invoke the model-specific unmarshaller.
+			err = unmarshaller(rawMap, rModelValue.Interface())
+			if err != nil {
+				err = fmt.Errorf(errorUnmarshalModel, propInsert(propertyName),
+					reflect.TypeOf(result).Elem().String(), err.Error())
+				return
+			}
+
+			// Add the model instance to the result slice (reflectively, of course :) ).
+			rResultSlice.Set(reflect.Append(rResultSlice, rModelValue.Elem().Elem()))
+		}
+	}
+	return
+}
+
+//
+// unmarshalModelMap unmarshals 'rawInput' into a map[string]<model>.
+//
+// Parameters:
+//
+// rawInput: the unmarshal input source in the form of a map[string]json.RawMessage.
+//
+// propertyName: an optional name of a property to be retrieved from 'rawInput'.
+// If 'propertyName' is specified as a non-empty string, then the specified property value is retrieved
+// from 'rawInput', and this value is assumed to contain the map[string]<model> to be unmarshalled.
+// If 'propertyName' is specified as "", then 'rawInput' will be used directly as the unmarshal input source.
+//
+// result: the unmarshal destination. This should be a pointer to map[string]<model> (e.g. *map[string]Foo for model type Foo).
+// If 'result' points to a nil map, this function will construct the map prior to adding entries to it.
+//
+// unmarshaller: the function used to unmarshal a single instance of the model.
+//
+func unmarshalModelMap(rawInput interface{}, propertyName string, result interface{}, unmarshaller ModelUnmarshaller) (err error) {
+	var rawMap map[string]json.RawMessage
+	var ok, foundIt bool
+
+	// rawInput should be a map[string]json.RawMessage.
+	rawMap, ok = rawInput.(map[string]json.RawMessage)
+	if !ok {
+		err = fmt.Errorf(errorIncorrectInputType, "map[string]json.RawMessage", reflect.TypeOf(rawInput).String())
+		return
+	}
+
+	// If propertyName was specified, then retrieve that entry from 'rawInput' as our unmarshal input source.
+	// Otherwise, just use 'rawInput' directly.
+	if propertyName != "" {
+		var rawMsg json.RawMessage
+		rawMsg, foundIt = rawMap[propertyName]
+		if foundIt && rawMsg != nil {
+			rawMap = make(map[string]json.RawMessage)
+			err = json.Unmarshal(rawMsg, &rawMap)
+			if err != nil {
+				err = fmt.Errorf(errorUnmarshalModel, propInsert(propertyName),
+					reflect.TypeOf(result).Elem().String(), err.Error())
+				return
+			}
+		}
+	} else {
+		foundIt = true
```

Comment:
Suggestion: set ``foundIt`` to ``true`` when you first initialize it so you can remove this extra condition. If ``rawMap`` doesn't have the property then it'll be false, otherwise it'll still be set to true